font-types/src/raw.rs

```diff
@@ -32,18 +41,23 @@ pub trait FixedSize: Sized {
     const RAW_BYTE_LEN: usize;
 }
 
-/// A trait for types that can be read from raw bytes.
-///
-/// This is a generalization that gives us a failable read method for all our
-/// `Scalar` types, as well as their `BigEndian` representations.
-///
-/// You should not need to implement this trait; it is provided automatically
-/// when you implement [`Scalar`].
-pub trait ReadScalar: FixedSize {
-    /// Interpret the provided bytes as `Self`, if they are the right length.
+/// we hide this trait; it isn't part of the public API, and this clarifies
+/// the guarantee that it is only implemented for [u8; N]
+mod sealed {
+    /// A trait representing any fixed-size big-endian byte array.
     ///
-    /// This should use all the provided bytes; bounds checking is performed upstream.
-    fn read(bytes: &[u8]) -> Option<Self>;
+    /// This is only used in `Scalar`, as a way of expressing the condition that the
+    /// `Raw` type is always a fixed-size byte array.
+    pub trait BeByteArray: Copy + AsRef<[u8]> {
```

Comment:
pub on something that isn't visible is weird particularly if you are newer to Rust. Would pub(crate) be ok?

Reply:
This is a Funny Rust Thing. `BeByteArray` needs to be `pub`, because it appears in another public type (the `Scalar` trait). However, we don't want anyone else to ever be able to implement this trait, because that would compromise our safety guarantees.

By placing the public type in the private module, it is still technically public, but can never be named, and thus cannot be implemented.

I was looking for a source for this pattern and I can't find exactly what I was looking for; the closest I can find is the [sealed traits](https://rust-lang.github.io/api-guidelines/future-proofing.html#sealed-traits-protect-against-downstream-implementations-c-sealed) pattern, which is similar but includes an extra trait.

Reply:
> By placing the public type in the private module, it is still technically public, but can never be named, and thus cannot be implemented.

This is very clever but perhaps not very obvious. Can you put a comment on it? - could have the exact words you used in your comment :)